### ISSUE TITLE
OneSignal Flutter v2.6.2 not working with Firebase Cloud Messaging 8.0.0-dev.14

### ISSUE THREAD

**Issue by davidbrenner** [OneSignal/OneSignal-Flutter-SDK#348]
**Description:**

OneSignal Flutter v2.6.2 not working with Firebase Cloud Messaging 8.0.0-dev.14

**Environment**

```
dependencies:
  flutter:
    sdk: flutter
  firebase_core: "0.7.0"
  firebase_analytics: "^7.0.1"
  firebase_auth: "^0.20.0+1"
  firebase_crashlytics: "^0.4.0+1"
  firebase_messaging: "^8.0.0-dev.14"
  cloud_firestore: "^0.16.0"
  firebase_dynamic_links: ^0.7.0+1
  firebase_in_app_messaging: "^0.3.0+1"
  firebase_storage: "^7.0.0"
  onesignal_flutter: ^2.6.2
```

```
% flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 1.22.6, on macOS 11.1 20C69 darwin-x64, locale en-US)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 12.4)
[!] Android Studio (version 4.1)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] VS Code (version 1.53.1)
[✓] Connected device (2 available)

! Doctor found issues in 1 category.
```

**Steps to Reproduce Issue:**

1. Start with project with working FCM
1. Install OneSiganal Flutter SDK following these instructions: https://documentation.onesignal.com/docs/flutter-sdk-setup
1. Attempt to receive a push notification to a physical iOS device sent via FCM console

The message is received, but the FCM handlers are not called when the message is opened. Before installing OneSignal, this was working. Additionally, FCM was working with OneSignal on older builds of this app. As part of tracking down this issue, I created a new project and added anything involving notification-related code step-by-step. Once OneSignal was added and configured, FCM's handling broke. 

We use both OneSignal and FCM APIs separately, but FCM is far more important to us than OneSignal, so our solution is going to be to just remove OneSignal. We would still like to use it if there is a solution.

**Anything else:**

N/A

**Comment by KieranLafferty** [comment]
I'm also experiencing this. The best solution to this would be to make swizzling optional. It's frustrating when frameworks implement swizzling and it breaks other dependencies and there is no way to turn it off

**Comment by emawby** [comment]
Hello can you try the steps listed [here](https://github.com/OneSignal/OneSignal-Flutter-SDK/issues/197#issuecomment-812660051)?

**Comment by davidbrenner** [comment]
We stopped using OneSignal because of this. If someone can confirm that solution works, I might be able to schedule some time into exploring this again, but for now I can't put more time into it.

**Comment by emawby** [comment]
Moving the steps into this issue as well for easier visibility.

If the application is subclassing FlutterAppDelegate (probably most common) the apps simply need to implement the UNUserNotificationCenter protocol in the `AppDelegate.h` file.
`@interface AppDelegate : FlutterAppDelegate <UNUserNotificationCenterDelegate>` 

set itself as the `UNUserNotificationCenterDelegate` in `application:didFinishLaunchingWithOptions:`
```
if (@available(iOS 10.0, *)) {
      [UNUserNotificationCenter currentNotificationCenter].delegate = (id<UNUserNotificationCenterDelegate>) self;
}
```

Finally in the protocol method implementations call Super. This calls the FlutterAppDelegate version of these methods.
```
- (void)userNotificationCenter:(UNUserNotificationCenter *)center willPresentNotification:(UNNotification *)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions))completionHandler {
    [super userNotificationCenter:center willPresentNotification:notification withCompletionHandler:completionHandler];
}

- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler {
    [super userNotificationCenter:center didReceiveNotificationResponse:response withCompletionHandler:completionHandler];
}

```

If your Application is not subclassing FlutterAppDelegate as described in the [Flutter doc here](https://flutter.dev/docs/development/add-to-app/ios/add-flutter-screen#using-the-flutterappdelegate) the app will do something similar but will call the method directly on the `_lifeCycleDelegate` instead of Super.
```
- (void)userNotificationCenter:(UNUserNotificationCenter *)center willPresentNotification:(UNNotification *)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions))completionHandler {
    [_lifeCycleDelegate userNotificationCenter:center willPresentNotification:notification withCompletionHandler:completionHandler];
}

- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler {
    [_lifeCycleDelegate userNotificationCenter:center didReceiveNotificationResponse:response withCompletionHandler:completionHandler];
}
```

**Comment by MohammedAlimoor** [comment]
@emawby   AppDelegate in swift

**Comment by emawby** [comment]
@MohammedAlimoor Your AppDelegate file may be in Swift or it maybe in AppDelegate.h and AppDelegate.m in Objective-C

**Comment by jelenalecic** [comment]
@emawby 
Can you please provide me with the whole AppDelegate.swift file?
Thank you!

**Comment by emawby** [comment]
@jelenalecic It doesn't look like your AppDelegate class specified that it is also a `UNUserNotificationCenterDelegate` on line 6

**Comment by emawby** [comment]
Closing this issue with the workaround posted in the thread.

**Comment by pictureframing** [comment]
Hi!
Unfortunately none of the solutions provided by @emawby ([here](https://github.com/OneSignal/OneSignal-Flutter-SDK/issues/348#issuecomment-824163607)) or @tayseer9 ([here](https://github.com/OneSignal/OneSignal-Flutter-SDK/issues/348#issuecomment-950626955)) worked for me.
Same problem as mentioned by @jelenalecic. The notification is received but not passed to dart.

When I try to specify that AppDelegate is also `UNUserNotificationCenterDelegate` like @emawby mentioned:
```
@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate, UNUserNotificationCenterDelegate {
...
}
```
Compilation fails due to the error:
```
Redundant conformance of 'AppDelegate' to protocol 'UNUserNotificationCenterDelegate'
```

I have tried every single approach I found in both the OneSignal and the Firebase Messaging GitHub repos. None of them worked. As long as OneSignal is specified as dependency in pubspec.yaml, Firebase Messaging won't work.

I will attach my current AppDelegate.swift (inspired by @tayseer9). Am I missing something?

<details>
    <summary>Flutter Doctor</summary>

    [✓] Flutter (Channel stable, 2.5.3, on macOS 11.5.2 20G95 darwin-x64, locale de-DE)
    [!] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
        ✗ cmdline-tools component is missing
        Run `path/to/sdkmanager --install "cmdline-tools;latest"`
        See https://developer.android.com/studio/command-line for more details.
        ✗ Android license status unknown.
        Run `flutter doctor --android-licenses` to accept the SDK licenses.
        See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.
    [✓] Xcode - develop for iOS and macOS
    [✓] Chrome - develop for the web
    [✓] Android Studio (version 3.6)
    [✓] VS Code (version 1.63.2)
    [✓] Connected device (2 available)
</details>

<details>
    <summary>AppDelegate.swift</summary>

    import UIKit
    import Flutter
    import FirebaseMessaging

    @UIApplicationMain
    @objc class AppDelegate: FlutterAppDelegate, UNUserNotificationCenterDelegate {
        override func application(
            _ application: UIApplication,
            didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
        ) -> Bool {
            GeneratedPluginRegistrant.register(with: self)

            if #available(iOS 10.0, *) {
                UNUserNotificationCenter.current().delegate = self as UNUserNotificationCenterDelegate
            }

            return super.application(application, didFinishLaunchingWithOptions: launchOptions)
        }

        override func application(_ application: UIApplication,
            didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
            Messaging.messaging().apnsToken = deviceToken;
        }
 
        override func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
            print("Push notification received on foreground");

            let userInfo = notification.request.content.userInfo
            Messaging.messaging().appDidReceiveMessage(userInfo)
            super.userNotificationCenter(center, willPresent: notification, withCompletionHandler: completionHandler)

            // Uncomment this line to receive notifications on foreground
            // completionHandler([[.alert, .sound]])
        }
    
        override func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
            print("Push notification received");

            super.userNotificationCenter(center, didReceive: response, withCompletionHandler: completionHandler)

            completionHandler()
        }
    }
</details>

**Comment by rserro** [comment]
@pictureframing Did you manage to solve it? I have the same problem.

**Comment by rserro** [comment]
@emawby I have exactly the same problem. If onesignal_flutter is in the pubspec, FirebaseMessaging handlers do not work. Tried everything in this thread. At end, I have the same "Redundant..." problem as @picturefarming

**Comment by pictureframing** [comment]
Hi @rserro, no we were not able to solve it. The answer from the customer support team of OneSignal was that it just does not work. No further help from their side. We postponed the project we needed this for but will probably go with another solution then having both of the packages as dependencies.

**Comment by rserro** [comment]
Ok, thanks @pictureframing !

**Comment by stevben** [comment]
Same problem for us...One Signal is not providing solutions and this is not a priority...very disappointing

**Comment by jkasten2** [comment]
Update, we were able to reproduce this bug and opened a bug report with FlutterFire. Feel free to monitor  https://github.com/firebase/flutterfire/issues/8792 for more updates.

**Comment by jkasten2** [comment]
Update, I have summited a PR to the FlutterFire SDK to accelerate the process in getting the issue fixed:
https://github.com/firebase/flutterfire/pull/8822

**Comment by jkasten2** [comment]
https://github.com/firebase/flutterfire/pull/8822 as been merged, this means it will be included in their next release. Closing this issue.

**Comment by jkasten2** [comment]
Google has shipped the fix for this in [firebase_messaging 11.4.2](https://pub.dev/packages/firebase_messaging/changelog#1142).
The following change log entry relates to this:
* FIX: Swizzle check for FlutterAppLifeCycleProvider instead of UNUserNotificationCenterDelegate (#8822). ([81f6b274](https://github.com/firebase/flutterfire/commit/81f6b2743b99e47c16fc3ee13cc1e7e6e7982730))
